### PR TITLE
[Feature]: Support Condinst

### PR DESCRIPTION
**Pull request by JosonChan1998**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add Condition instance segmentation methods. The results are listed below:
MS_R_50_1x| box AP | mask AP |
--- |:---:|:---:|
origin   | 39.7 | 35.7 | 
ours| 39.6 | 35.2 |

The reproduction results are still a little behind, probably because there are more iterations in detectron2.

## Modification

1. configs/condinst
2. mmdet/models/detectors/condinst.py
3. mmdet/models/dense_heads/condinst_head.py
4. mmdet/models/utils/misc.py

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
5. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
6. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
7. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by JosonChan1998** [comment]
@RangiLyu  Hi~ Any progress?

File: mmdet/models/dense_heads/condinst_head.py [hunk 1]
```diff
@@ -0,0 +1,1185 @@
...
+
+        return self._bbox_post_process(
+            results=results,
+            cfg=cfg,
+            rescale=rescale,
+            with_nms=with_nms,
+            img_meta=img_meta)
+
+
+class CondInstMaskFeatModule(MaskFeatModule):
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/condinst_head.py:711]: All of the methods in SOLOv2 MaskFeatModule are overwritten in this class. So the inheritance seems unnecessary.

**Review by RangiLyu — COMMENTED** [review]
> The reproduction results are still a little behind, probably because there are more iterations in detectron2.

Is the precision of the inference aligned? Can this implementation get the same box and mask AP with a converted weight from the original version?

File: configs/condinst/condinst_r50_fpn_ms_1x_coco.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
+_base_ = [
```
Comment (RangiLyu) [inline configs/condinst/condinst_r50_fpn_ms_1x_coco.py:1]: You can also add a config based on the detectron2's 90k schedule in configs/common/ms-poly-90k_coco-instance.py

File: mmdet/models/dense_heads/condinst_head.py [hunk 3]
```diff
@@ -0,0 +1,1185 @@
...
+        """Compute regression and classification targets for a single image."""
+        num_points = points.size(0)
+        num_gts = len(gt_instances)
+        gt_bboxes = gt_instances.bboxes
+        gt_labels = gt_instances.labels
+        gt_masks = gt_instances.get('masks', None)
+
+        if num_gts == 0:
+            return gt_labels.new_full((num_points,), self.num_classes), \
+                   gt_bboxes.new_zeros((num_points, 4))
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/condinst_head.py:289]: The length of the return values are mismatched, which will cause an error when `num_gts == 0`

File: mmdet/models/dense_heads/condinst_head.py [hunk 4]
```diff
@@ -0,0 +1,1185 @@
...
+                yys = torch.arange(
+                    0, h, dtype=torch.float32, device=masks.device)
+                xxs = torch.arange(
+                    0, w, dtype=torch.float32, device=masks.device)
+
+                m00 = masks.sum(dim=-1).sum(dim=-1).clamp(min=1e-6)
+                m10 = (masks * xxs).sum(dim=-1).sum(dim=-1)
+                m01 = (masks * yys[:, None]).sum(dim=-1).sum(dim=-1)
+                center_xs = m10 / m00
+                center_ys = m01 / m00
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/condinst_head.py:328]: Add some comments about the variable names m00, m01...

File: mmdet/models/dense_heads/condinst_head.py [hunk 5]
```diff
@@ -0,0 +1,1185 @@
...
+        pos_inds = ((labels >= 0)
+                    & (labels < bg_class_ind)).nonzero().reshape(-1)
+        pos_gt_inds = min_area_inds[labels < self.num_classes]
+        return labels, bbox_targets, pos_inds, pos_gt_inds
+
+    def get_positive_infos(self) -> InstanceList:
+        """Get positive information from sampling results.
+
+        Returns:
+            list[:obj:`InstanceData`]: Positive information of each image,
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/condinst_head.py:393]: If the return value can be None, please change the type-hint and docstring,

File: mmdet/models/dense_heads/condinst_head.py [hunk 6]
```diff
@@ -0,0 +1,1185 @@
...
+                  shape (num_instances, h, w).
+        """
+        cfg = self.test_cfg if cfg is None else cfg
+        scale_factor = bboxes.new_tensor(img_meta['scale_factor']).repeat(
+            (1, 2))
+        img_h, img_w = img_meta['img_shape'][:2]
+        ori_h, ori_w = img_meta['ori_shape'][:2]
+
+        mask_preds = mask_preds.sigmoid().unsqueeze(0)
+        if rescale:  # in-placed rescale the bboxes
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/condinst_head.py:1173]: "rescale" means rescale to the origin image shape.

File: mmdet/models/dense_heads/condinst_head.py [hunk 7]
```diff
@@ -0,0 +1,1185 @@
...
+            scale_factor = bboxes.new_tensor(img_meta['scale_factor']).repeat(
+                (1, 2))
+            bboxes /= scale_factor
+
+            mask_preds = aligned_bilinear(mask_preds, self.mask_out_stride)
+            mask_preds = mask_preds[:, :, :img_h, :img_w]
+
+        masks = F.interpolate(
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/condinst_head.py:1181]: Put this interpolation into the `rescale`

File: mmdet/models/utils/misc.py [hunk 8]
```diff
@@ -560,3 +561,86 @@ def reweight_loss_dict(losses: dict, weight: float) -> dict:
             else:
                 losses[name] = loss * weight
     return losses
+
+
+def compute_locations(h: int, w: int, feat_stride: int, device: str) -> Tensor:
```
Comment (RangiLyu) [inline mmdet/models/utils/misc.py:566]: Can this be replaced with `MlvlPointGenerator.single_level_grid_priors`?

File: mmdet/models/dense_heads/condinst_head.py [hunk 9]
```diff
@@ -0,0 +1,1185 @@
...
+                                                        self.in_channels)
+            else:
+                # out_channels x in_channels x 1 x 1
+                weight_splits[i] = weight_splits[i].reshape(
+                    num_insts * 1, -1, 1, 1)
+                bias_splits[i] = bias_splits[i].reshape(num_insts)
+
+        return weight_splits, bias_splits
+
+    def mask_heads_forward(self, features: Tensor, weights: List[Tensor],
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/condinst_head.py:892]: dynamic_conv_forward would be better

File: configs/condinst/condinst_r50_fpn_ms_1x_coco.py (hunk 2 above)
Comment (JosonChan1998) [reply configs/condinst/condinst_r50_fpn_ms_1x_coco.py]: Thanks for your kind suggestion. I have tried the 90k shcedule and i got a precision:
MS_R_50_1x| box AP | mask AP |
--- |:---:|:---:|
origin   | 39.7 | 35.7 | 
ours| 39.9 | 35.5 |

Is a AP error of 0.2% acceptable?

Comment (JosonChan1998) [reply configs/condinst/condinst_r50_fpn_ms_1x_coco.py]: > Great! I think an error of 0.2% is acceptable.

Thanks! 
I have made corrections based on your comments, can you please review it again?

**Comment by hhaAndroid** [comment]
@JosonChan1998 I think this PR is very well. Just need to add unit tests, I have no other comments.

File: mmdet/models/dense_heads/condinst_head.py [hunk 10]
```diff
@@ -0,0 +1,1225 @@
...
+        self.start_level = start_level
+        self.end_level = end_level
+        self.mask_stride = mask_stride
+        self.num_stacked_convs = num_stacked_convs
+        assert start_level >= 0 and end_level >= start_level
+        self.out_channels = out_channels
+        self.conv_cfg = conv_cfg
+        self.norm_cfg = norm_cfg
+        self._init_layers()
+        self.fp16_enabled = False
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/condinst_head.py:764]: delete

File: mmdet/models/dense_heads/condinst_head.py [hunk 11]
```diff
@@ -0,0 +1,1225 @@
...
+                    padding=1,
+                    conv_cfg=self.conv_cfg,
+                    norm_cfg=self.norm_cfg,
+                    bias=False))
+        self.conv_branch = nn.Sequential(*conv_branch)
+
+        self.conv_pred = nn.Conv2d(
+            self.feat_channels, self.out_channels, 1, stride=1)
+
+    def init_weights(self) -> None:
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/condinst_head.py:800]: init_weights and init_cfg are not compatible, if you want both to take effect, you need to change it to

```python
super().init_weights()
kaiming_init(self.convs_all_levels, a=1, distribution='uniform')
...
```

File: configs/condinst/condinst_r50_fpn_ms_90k_coco_instance.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
+_base_ = '../common/ms-poly-90k_coco-instance.py'
+
+train_pipeline = [
+    dict(
+        type='LoadImageFromFile',
+        file_client_args={{_base_.file_client_args}},
+        imdecode_backend={{_base_.backend}}),
+    dict(
+        type='LoadAnnotations', with_bbox=True, with_mask=True,
+        poly2mask=True),
```
[the remark this answers is not in the document]
Comment (JosonChan1998) [reply configs/condinst/condinst_r50_fpn_ms_90k_coco_instance.py:10]: If poly2mask is False, then we should change poly to mask in loss compute and target assign. I think it will slow down the speed.

File: mmdet/models/dense_heads/condinst_head.py (hunk 10 above)
Comment (JosonChan1998) [reply mmdet/models/dense_heads/condinst_head.py:764]: Is the `fp16_enabled` should be deleted?

Comment (RangiLyu) [reply mmdet/models/dense_heads/condinst_head.py]: Yes. It is no longer needed in mmdet 3.x.

Comment (JosonChan1998) [reply mmdet/models/dense_heads/condinst_head.py]: ok! I have fixed it.

File: mmdet/models/dense_heads/condinst_head.py [hunk 13]
```diff
@@ -0,0 +1,1225 @@
...
+            scale_factor = bboxes.new_tensor(img_meta['scale_factor']).repeat(
+                (1, 2))
+            bboxes /= scale_factor
+
+            masks = F.interpolate(
+                mask_preds, (ori_h, ori_w),
+                mode='bilinear',
+                align_corners=False).squeeze(0) > cfg.mask_thr
+        else:
+            masks = mask_preds > cfg.mask_thr
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/condinst_head.py:1223]: You also need to add `squeeze(0)` here. Or you can move the `unsqueeze(0)` into the rescale part.

File: configs/condinst/condinst_r50_fpn_ms_90k_coco_instance.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (RangiLyu) [reply configs/condinst/condinst_r50_fpn_ms_90k_coco_instance.py:10]: The RandomChoiceResize for polygon masks is much faster than bitmap masks and cost less memory.

**Review by RangiLyu — COMMENTED** [review]
The code looks good to me now. Would you like to write the README.md file?

**Review by RangiLyu — COMMENTED** [review]
Remember to use polygon masks. And after this change, the config file name should be renamed to `condinst_r50_fpn_ms-poly-90k_coco.py`

**Comment by JosonChan1998** [comment]
> Remember to use polygon masks. And after this change, the config file name should be renamed to `condinst_r50_fpn_ms-poly-90k_coco.py`

Done.

**Review by RangiLyu — DISMISSED** [review]

**Review by RangiLyu — APPROVED** [review]
